sstables/sstables.cc

```diff
@@ -1939,6 +1939,32 @@ bool sstable::has_component(component_type f) const {
     return _recognized_components.contains(f);
 }
 
+std::optional<bool> sstable::originated_on_this_node() const {
+    if (_version < version_types::me) {
+        // earlier formats do not store originating host id
+        return std::nullopt;
+    }
+
+    auto originating_host_id = get_stats_metadata().originating_host_id;
+    if (!originating_host_id) {
+        // Scylla always fills in originating host id when writing
+        // sstables, so an ME-and-up sstable that does not have it is
+        // invalid
+        on_internal_error(sstlog, format("No originating host id in SSTable: {}. Load foreign SSTables via the upload dir instead.", get_filename()));
```

Comment:
Do we know that load-n-stream or refresh won't move sstables in? Will it always rewrite them?

Reply:
btw on_internal_error is for internal errors, not reporting user errors.

Reply:
which means, in turn, that we must do this check when loading the sstables in the first place.

Reply:
> Do we know that load-n-stream or refresh won't move sstables in? Will it always rewrite them?

Yes, we always rewrite (reshard even) in refresh.
And apply mutations with load-and-stream.

> btw on_internal_error is for internal errors, not reporting user errors.

What would be the user error? Placing a foreign sstable directly into the table's directory?
If the sstable has scylla_metadata it would be an internal error by the way.
I can degrade it to a regular log error if you think it's more appropriate.

> which means, in turn, that we must do this check when loading the sstables in the first place.

We do. It's called from `sstable_directory::validate` in the `!sst->is_uploaded()` case.

Reply:
> What would be the user error? Placing a foreign sstable directly into the table's directory?

We do this in the field for schema sstables when restoring schema or working around schema sync issues.

Reply:
Yes, but they will have a host_id if they are in me format. By "foreign" I meant from another vendor, not another node.